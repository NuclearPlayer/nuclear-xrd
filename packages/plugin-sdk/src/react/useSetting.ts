import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
