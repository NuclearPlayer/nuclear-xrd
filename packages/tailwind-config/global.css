@import 'tailwindcss';
@import 'tw-animate-css';

@source "../player";
@source "../ui";
@source "../storybook";

@theme {
  --color-background: var(--background);
  --color-background-secondary: var(--background-secondary);
  --color-background-input: var(--background-input);

  --color-foreground: var(--foreground);
  --color-foreground-secondary: var(--foreground-secondary);
  --color-foreground-input: var(--foreground-input);

  --color-primary: var(--primary);
  --color-primary-hover: var(--primary-hover);
  --color-primary-active: var(--primary-active);

  --color-border: var(--border);
  --color-border-input: var(--border-input);
  --color-ring: var(--ring);

  --color-accent-green: var(--accent-green);
  --color-accent-yellow: var(--accent-yellow);
  --color-accent-purple: var(--accent-purple);
  --color-accent-blue: var(--accent-blue);
  --color-accent-orange: var(--accent-orange);
  --color-accent-cyan: var(--accent-cyan);
  --color-accent-red: var(--accent-red);

  --color-shadow: var(--shadow-color);

  --radius: var(--radius);
  --radius-sm: var(--radius-sm);
  --radius-lg: var(--radius-lg);

  --font-weight-normal: var(--font-weight-normal);
  --font-weight-bold: var(--font-weight-bold);

  --font-sans: var(--font-family);

  --shadow: var(--shadow-x) var(--shadow-y) 0px 0px var(--border);
  --shadow-shadow: var(--shadow);
  --shadow-nav: 4px 4px 0px 0px var(--border);
  --shadow-darkNav: 4px 4px 0px 0px var(--dark-border);
  --spacing-shadow-x: var(--shadow-x);
  --spacing-shadow-y: var(--shadow-y);
}

@custom-variant dark (&:where([data-theme=dark], [data-theme=dark] *));

:root {
  /* Core Backgrounds */
  --background: oklch(0.9491 0.023 341.75);
  --background-secondary: oklch(100% 0 0);
  --background-input: oklch(100% 0 0);

  /* Text Colors */
  --foreground: oklch(0% 0 0);
  --foreground-secondary: oklch(0.3 0 0);
  --foreground-input: oklch(0% 0 0);

  /* Interactive Elements */
  --primary: oklch(76.91% 0.173 341.75);
  --primary-hover: oklch(0% 0 0);
  --primary-active: oklch(0% 0 0);

  /* Borders & Outlines */
  --border: oklch(0% 0 0);
  --ring: oklch(100% 0 0);

  /* Accent Colors */
  --accent-green: oklch(79.05% 0.209 147.58);
  --accent-yellow: oklch(95.53% 0.134 112.76);
  --accent-purple: oklch(74.2% 0.149 301.88);
  --accent-blue: oklch(55.98% 0.08 270.09);
  --accent-orange: oklch(83.39% 0.124 66.56);
  --accent-cyan: oklch(88.26% 0.093 212.85);
  --accent-red: oklch(68.22% 0.206 24.43);

  /* Shadows */
  --shadow-color: oklch(0% 0 0);
  --shadow-x: 2px;
  --shadow-y: 2px;
  --shadow-blur: 0px;

  /* Typography */
  --font-family: 'DM Sans', system-ui, -apple-system, sans-serif;
  --font-weight-normal: 400;
  --font-weight-bold: 700;

  /* Border Radius */
  --radius: 8px;
  --radius-sm: 4px;
  --radius-lg: 12px;
}

@layer base {
  body {
    @apply text-foreground bg-background font-normal;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    @apply font-bold;
  }
}

[data-theme='dark'] {
  --background: oklch(0.4408 0.0838 341.75);
  --background-secondary: oklch(39.49% 0.031 277.49);
  --background-input: oklch(0% 0 0);

  --foreground: oklch(0.9249 0 0);
  --foreground-secondary: oklch(0.9249 0 0);
  --foreground-input: oklch(100% 0 0);
}
