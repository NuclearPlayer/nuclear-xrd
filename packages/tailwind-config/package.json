{
  "name": "@nuclearplayer/tailwind-config",
  "version": "0.0.7",
  "description": "Shared Tailwind CSS configuration for Nuclear",
  "type": "module",
  "main": "./global.css",
  "exports": {
    ".": "./global.css"
  },
  "files": [
    "global.css"
  ],
  "scripts": {
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
  },
  "dependencies": {
    "@tailwindcss/vite": "^4.1.11",
    "tailwindcss": "^4.1.11",
    "tw-animate-css": "^1.3.6"
  },
  "devDependencies": {
    "@nuclearplayer/eslint-config": "workspace:*"
  }
}
