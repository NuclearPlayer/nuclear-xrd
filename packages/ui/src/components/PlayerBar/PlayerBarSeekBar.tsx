import { FC } from 'react';

import '../../styles.css';

import { cn } from '../../utils';
import { formatTimeSeconds } from '../../utils/time';
import { useSeekBar } from './useSeekBar';

export type PlayerSeekBarProps = {
  progress: number;
  elapsedSeconds: number;
  remainingSeconds: number;
  isLoading?: boolean;
  onSeek?: (percent: number) => void;
  className?: string;
};

export const PlayerBarSeekBar: FC<PlayerSeekBarProps> = ({
  progress,
  elapsedSeconds,
  remainingSeconds,
  isLoading = false,
  onSeek,
  className = '',
}) => {
  const { clamped, containerRef, handleClick, isInteractive } = useSeekBar({
    progress,
    isLoading,
    onSeek,
  });

  return (
    <div className={cn('w-full px-4 select-none', className)}>
      <div className="mb-1 flex items-center justify-between text-xs leading-none">
        <span className="text-foreground-secondary tabular-nums">
          {formatTimeSeconds(elapsedSeconds)}
        </span>
        <span className="text-foreground-secondary tabular-nums">
          {formatTimeSeconds(-Math.abs(remainingSeconds))}
        </span>
      </div>
      <div
        ref={containerRef}
        className={cn('relative h-2 w-full', {
          'pointer-events-none cursor-not-allowed': isLoading,
          'cursor-pointer': isInteractive,
        })}
        onClick={handleClick}
        aria-disabled={isLoading}
      >
        <div
          className={cn(
            'border-border bg-background-secondary shadow-shadow absolute inset-0 border',
            { 'overflow-hidden': isLoading },
          )}
        >
          {isLoading && (
            <div className="bg-stripes-diagonal absolute inset-0 opacity-80" />
          )}
          {!isLoading && (
            <div
              className={cn('bg-primary h-full', 'transition-none')}
              style={{ width: `${clamped}%` }}
            />
          )}
        </div>
      </div>
    </div>
  );
};
