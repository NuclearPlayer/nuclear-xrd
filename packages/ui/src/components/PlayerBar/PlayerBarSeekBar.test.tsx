import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import '@nuclearplayer/tailwind-config';

import { PlayerBarSeekBar } from './PlayerBarSeekBar';

describe('PlayerBarSeekBar', () => {
  it('(Snapshot) default render', () => {
    const { asFragment } = render(
      <PlayerBarSeekBar
        progress={25}
        elapsedSeconds={97}
        remainingSeconds={1297}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onSeek with percent based on click position', () => {
    const onSeek = vi.fn();
    render(
      <div style={{ width: 200 }}>
        <PlayerBarSeekBar
          progress={0}
          elapsedSeconds={0}
          remainingSeconds={0}
          onSeek={onSeek}
        />
      </div>,
    );
    const container = document.querySelector('[aria-disabled]') as HTMLElement;
    const rect = {
      left: 0,
      width: 200,
      top: 0,
      height: 8,
      right: 200,
      bottom: 8,
    };
    vi.spyOn(container, 'getBoundingClientRect').mockReturnValue(
      rect as DOMRect,
    );
    fireEvent.click(container, { clientX: 50 });
    expect(onSeek).toHaveBeenCalledWith(25);
    fireEvent.click(container, { clientX: 200 });
    expect(onSeek).toHaveBeenCalledWith(100);
  });

  it('does not call onSeek when loading', () => {
    const onSeek = vi.fn();
    render(
      <div style={{ width: 200 }}>
        <PlayerBarSeekBar
          progress={50}
          elapsedSeconds={0}
          remainingSeconds={0}
          isLoading
          onSeek={onSeek}
        />
      </div>,
    );
    const container = document.querySelector(
      '[aria-disabled="true"]',
    ) as HTMLElement;
    const rect = {
      left: 0,
      width: 200,
      top: 0,
      height: 8,
      right: 200,
      bottom: 8,
    };
    vi.spyOn(container, 'getBoundingClientRect').mockReturnValue(
      rect as DOMRect,
    );
    fireEvent.click(container, { clientX: 100 });
    expect(onSeek).not.toHaveBeenCalled();
  });
});
