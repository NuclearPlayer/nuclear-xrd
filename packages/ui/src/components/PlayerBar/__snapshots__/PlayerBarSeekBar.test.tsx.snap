// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerBarSeekBar > (Snapshot) default render 1`] = `
<DocumentFragment>
  <div
    class="w-full px-4 select-none"
  >
    <div
      class="mb-1 flex items-center justify-between text-xs leading-none"
    >
      <span
        class="text-foreground-secondary tabular-nums"
      >
        1:37
      </span>
      <span
        class="text-foreground-secondary tabular-nums"
      >
        -21:37
      </span>
    </div>
    <div
      aria-disabled="false"
      class="relative h-2 w-full"
    >
      <div
        class="border-border bg-background-secondary shadow-shadow absolute inset-0 border"
      >
        <div
          class="bg-primary h-full transition-none"
          style="width: 25%;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
